rllib/utils/numpy.py

```diff
@@ -270,9 +272,87 @@ def lstm(x,
+
+    Returns a single 1D tensor resulting from the different input
+    components' values.
+
+    Thereby:
+    - Boxes (any shape) get flattened to (B, [T]?, -1). Note that image boxes
+    are not treated differently from other types of Boxes and get
+    flattened as well.
+    - Discrete (int) values are one-hot'd, e.g. a batch of [1, 0, 3] (B=3 with
+    Discrete(4) space) results in [[0, 1, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1]].
+    - MultiDiscrete values are multi-one-hot'd, e.g. a batch of
+    [[0, 2], [1, 4]] (B=2 with MultiDiscrete([2, 5]) space) results in
+    [[1, 0,  0, 0, 1, 0, 0], [0, 1,  0, 0, 0, 0, 1]].
+
+    Args:
+        inputs: The inputs to be flattened.
+        spaces_struct: The structure of the spaces that behind the input
+        time_axis: Whether all inputs have a time-axis (after the batch axis).
+            If True, will keep not only the batch axis (0th), but the time axis
+            (1st) as-is and flatten everything from the 2nd axis up.
+
+    Returns:
+        A single 1D tensor resulting from concatenating all
+        flattened/one-hot'd input components. Depending on the time_axis flag,
+        the shape is (B, n) or (B, T, n).
+
+    Examples:
+        >>> # B=2
+        >>> out = flatten_inputs_to_1d_tensor(
+        ...     {"a": [1, 0], "b": [[[0.0], [0.1]], [1.0], [1.1]]},
+        ...     spaces_struct=dict(a=Discrete(2), b=Box(shape=(2, 1)))
+        ... )
+        >>> print(out)
+        ... [[0.0, 1.0,  0.0, 0.1], [1.0, 0.0,  1.0, 1.1]]  # B=2 n=4
+
+        >>> # B=2; T=2
+        >>> out = flatten_inputs_to_1d_tensor(
+        ...     ([[1, 0], [0, 1]],
+        ...      [[[0.0, 0.1], [1.0, 1.1]], [[2.0, 2.1], [3.0, 3.1]]]),
+        ...     spaces_struct=tuple([Discrete(2), Box(shape=(2, ))]),
+        ...     time_axis=True
+        ... )
+        >>> print(out)
+        ... [[[0.0, 1.0, 0.0, 0.1], [1.0, 0.0, 1.0, 1.1]],
+        ...  [[1.0, 0.0, 2.0, 2.1], [0.0, 1.0, 3.0, 3.1]]]  # B=2 T=2 n=4
+    """
+
+    flat_inputs = tree.flatten(inputs)
+    flat_spaces = tree.flatten(spaces_struct) if spaces_struct is not None \
+        else [None] * len(flat_inputs)
+
+    out = []
+    for input_, space in zip(flat_inputs, flat_spaces):
+        assert isinstance(input_, np.ndarray)
+        # One-hot encoding.
+        if isinstance(space, (Discrete, MultiDiscrete)):
+            if isinstance(space, Discrete):
+                out.append(one_hot(input_, depth=space.n).astype(np.float32))
+            elif isinstance(space, MultiDiscrete):
+                out.append(
+                    np.concatenate(
+                        [
+                            one_hot(input_[:, i], depth=n).astype(np.float32)
+                            for i, n in enumerate(space.nvec)
+                        ],
+                        axis=-1))
+        # Flatten.
+        else:
+            out.append(
+                input_.reshape([input_.shape[0], -1]).astype(np.float32))
```

Comment:
minor nit: maybe define B = input_.shape[0], then do

input_.reshape([B, -1]).astype(np.float32)

I found it pretty useful to understand the intended shape here, and consistent with the other util funcs.